python/ray/data/_internal/stats.py

```diff
@@ -275,137 +332,304 @@ def summary_string(
+                stage_name = stage_stats_summary.stage_name
                 stage_uuid = self.dataset_uuid + stage_name
                 out += "\n"
                 out += "\tSubstage {} {}: ".format(n, stage_name)
                 if stage_uuid in already_printed:
                     out += "\t[execution cached]\n"
                 else:
                     already_printed.add(stage_uuid)
-                    out += self._summarize_blocks(metadata, is_substage=True)
-        out += self._summarize_iter()
+                    out += str(stage_stats_summary)
+        out += str(self.iter_stats)
+        if self.extra_metrics:
+            indent = "\t" if stage_stats_summary.is_substage else ""
+            out += indent
+            out += "* Extra metrics: " + str(self.extra_metrics) + "\n"
         return out
 
-    def _summarize_iter(self) -> str:
-        out = ""
-        if (
-            self.iter_total_s.get()
-            or self.iter_wait_s.get()
-            or self.iter_next_batch_s.get()
-            or self.iter_format_batch_s.get()
-            or self.iter_get_s.get()
-        ):
-            out += "\nDataset iterator time breakdown:\n"
-            out += "* In ray.wait(): {}\n".format(fmt(self.iter_wait_s.get()))
-            out += "* In ray.get(): {}\n".format(fmt(self.iter_get_s.get()))
-            out += "* In next_batch(): {}\n".format(fmt(self.iter_next_batch_s.get()))
-            out += "* In format_batch(): {}\n".format(
-                fmt(self.iter_format_batch_s.get())
-            )
-            out += "* In user code: {}\n".format(fmt(self.iter_user_s.get()))
-            out += "* Total time: {}\n".format(fmt(self.iter_total_s.get()))
-        return out
+    def get_total_wall_time(self) -> float:
+        parent_wall_times = [p.get_total_wall_time() for p in self.parents]
+        parent_max_wall_time = max(parent_wall_times) if parent_wall_times else 0
+        return parent_max_wall_time + sum(
+            ss.wall_time.get("max", 0) for ss in self.stages_stats
+        )
+
+    def get_total_cpu_time(self) -> float:
+        parent_sum = sum(p.get_total_cpu_time() for p in self.parents)
+        return parent_sum + sum(ss.cpu_time.get("sum", 0) for ss in self.stages_stats)
+
+    def get_max_heap_memory(self) -> float:
+        parent_memory = [p.get_max_heap_memory() for p in self.parents]
+        parent_max = max(parent_memory) if parent_memory else 0
+        return max(
+            parent_max,
+            *[ss.memory.get("max", 0) for ss in self.stages_stats],
+        )
+
+
+@dataclass
+class StageStatsSummary:
+    stage_name: str
+    # Whether the stage associated with this StageStatsSummary object is a substage
+    is_substage: bool
+    # Total stage walltime
+    time_total_s: float
+    # String summarizing high-level statistics from executing the stage
+    block_execution_summary_str: str
+    # Dict with aggregated stats of wall times:
```

Comment:
nit: since the fields are mostly following the same semantics, it's easier to write this just once. It's good to call out those min/max/sum/mean are calculated across blocks processed in the stage.